File: iOS/TUIKit/TUICallKit/TUICallKit-Swift/ViewModel/Component/VideoLayout/GroupCallVideoLayoutViewModel.swift
//
//  GroupCallVideoLayoutViewModel.swift
//  TUICallKit
//
//  Created by vincepzhang on 2023/3/7.
//

import Foundation
import TUICallEngine

class GroupCallVideoLayoutViewModel {
    
    let callStatusObserver = Observer()
    let mediaTypeObserver = Observer()
    let remoteUserListObserver = Observer()
    let callRoleCountObserver = Observer()
    let isCameraOpenObserver = Observer()
    
    let selfUser: Observable<User> = Observable(User())
    let selfCallStatus: Observable<TUICallStatus> = Observable(.none)
    let selfCallRole: Observable<TUICallRole> = Observable(.none)
    let mediaType: Observable<TUICallMediaType> = Observable(.unknown)
    let remoteUserList: Observable<[User]> = Observable(Array())
    let isCameraOpen: Observable<Bool> = Observable(false)
    
    var allUserList = [User]()
    
    init() {
        selfUser.value = TUICallState.instance.selfUser.value
        selfCallStatus.value = TUICallState.instance.selfUser.value.callStatus.value
        selfCallRole.value = TUICallState.instance.selfUser.value.callRole.value
        mediaType.value = TUICallState.instance.mediaType.value
        remoteUserList.value = TUICallState.instance.remoteUserList.value
        isCameraOpen.value = TUICallState.instance.isCameraOpen.value
        processUserList(remoteUserList: remoteUserList.value)
        registerObserver()
    }
    
    deinit {
        TUICallState.instance.selfUser.value.callRole.removeObserver(callRoleCountObserver)
        TUICallState.instance.selfUser.value.callStatus.removeObserver(callStatusObserver)
        TUICallState.instance.mediaType.removeObserver(mediaTypeObserver)
        TUICallState.instance.remoteUserList.removeObserver(remoteUserListObserver)
        TUICallState.instance.isCameraOpen.removeObserver(isCameraOpenObserver)
    }
    
    func registerObserver() {
        TUICallState.instance.selfUser.value.callRole.addObserver(callRoleCountObserver, closure: { [weak self] newValue, _ in
            guard let self = self else { return }
            self.selfCallRole.value = newValue
        })
        
        TUICallState.instance.selfUser.value.callStatus.addObserver(callStatusObserver, closure: { [weak self] newValue, _ in
            guard let self = self else { return }
            self.selfCallStatus.value = newValue
        })
        
        TUICallState.instance.mediaType.addObserver(mediaTypeObserver, closure: { [weak self] newValue, _ in
            guard let self = self else { return }
            self.mediaType.value = newValue
        })
        
        TUICallState.instance.remoteUserList.addObserver(remoteUserListObserver, closure: { [weak self] newValue, _ in
            guard let self = self else { return }
            self.processUserList(remoteUserList: newValue)
            self.remoteUserList.value = newValue
        })
        
        TUICallState.instance.isCameraOpen.addObserver(isCameraOpenObserver, closure: { [weak self] newValue, _ in
            guard let self = self else { return }
            self.isCameraOpen.value = newValue
        })
    }
    
    func processUserList(remoteUserList: [User]) {
        allUserList.removeAll()
        selfUser.value.index = 0
        allUserList.append(selfUser.value)
        
        for (index, value) in remoteUserList.enumerated() {
            value.index = index + 1
            allUserList.append(value)
        }
    }
    
    // MARK: Set TUICallState showLargeViewUserId
    func setShowLargeViewUserId(userId: String) {
        TUICallState.instance.showLargeViewUserId.value = userId
    }
    
}

File: Android/TUIKit/TUICallKit/tuicallkit-kt/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/viewmodel/component/videolayout/GroupCallVideoLayoutViewModel.kt
package com.tencent.qcloud.tuikit.tuicallkit.viewmodel.component.videolayout

import com.tencent.qcloud.tuikit.TUICommonDefine
import com.tencent.qcloud.tuikit.tuicallengine.TUICallDefine
import com.tencent.qcloud.tuikit.tuicallengine.impl.base.LiveData
import com.tencent.qcloud.tuikit.tuicallengine.impl.base.Observer
import com.tencent.qcloud.tuikit.tuicallkit.data.User
import com.tencent.qcloud.tuikit.tuicallkit.state.TUICallState
import java.util.concurrent.CopyOnWriteArrayList

class GroupCallVideoLayoutViewModel {
    public var userList = LiveData<CopyOnWriteArrayList<User>>()
    public var isCameraOpen = LiveData<Boolean>()
    public var isFrontCamera = LiveData<TUICommonDefine.Camera>()
    public var mediaType = LiveData<TUICallDefine.MediaType>()
    public var changedUser = LiveData<User>()
    public var showLargeViewUserId = LiveData<String>()

    private var remoteUserListObserver = Observer<LinkedHashSet<User>> {
        if (it != null && it.size > 0) {
            for (user in it) {
                if (!userList.get().contains(user)) {
                    userList.get().add(user)
                    changedUser.set(user)
                }
            }
            for ((index, user) in userList.get().withIndex()) {
                if (index == 0) {
                    continue
                }
                if (!it.contains(user)) {
                    user.callStatus.set(TUICallDefine.Status.None)
                    userList.get().remove(user)
                    changedUser.set(user)
                }
            }
        }
    }

    init {
        isCameraOpen = TUICallState.instance.isCameraOpen
        isFrontCamera = TUICallState.instance.isFrontCamera
        mediaType = TUICallState.instance.mediaType
        changedUser.set(null)
        userList.set(CopyOnWriteArrayList())
        userList.get().add(TUICallState.instance.selfUser.get())
        userList.get().addAll(TUICallState.instance.remoteUserList.get())
        showLargeViewUserId = TUICallState.instance.showLargeViewUserId

        addObserver()
    }

    private fun addObserver() {
        TUICallState.instance.remoteUserList.observe(remoteUserListObserver)
    }

    public fun removeObserver() {
        TUICallState.instance.remoteUserList.removeObserver(remoteUserListObserver)
    }

    fun updateShowLargeViewUserId(userId: String?) {
        TUICallState.instance.showLargeViewUserId.set(userId)
    }

    fun updateBottomViewExpanded(isExpand: Boolean) {
        if (isExpand != TUICallState.instance.isBottomViewExpand.get()) {
            TUICallState.instance.isBottomViewExpand.set(!TUICallState.instance.isBottomViewExpand.get())
        }
    }
}